File: src/main/java/Version_Java/West2FriedChickenRestaurant.java
package Version_Java;

import java.time.LocalDate;
import java.util.LinkedList;

public class West2FriedChickenRestaurant implements FriedChickenRestaurant{

    private double amount;
    private final LinkedList<Beer> beers = new LinkedList<Beer>();
    private final LinkedList<Juice> juices = new LinkedList<Juice>();
    private final LinkedList<SetMeal> meals = new LinkedList<SetMeal>();
    // Answer:
    // 使用 LinkedList 也就是 链表结构 添加 和 删除 更为高效

    public West2FriedChickenRestaurant() {
            amount = 12; // 初始资金
            meals.add(
                    new SetMeal("SuperChicken",70.0,"WhiteChicken",
                            new Juice("Snow",4.0,
                                    LocalDate.of(2020,11,1)))
            );
            meals.add(
                new SetMeal("NiceChicken",50.0,"GreenChicken",
                        new Beer("Sun",4.0,
                                LocalDate.of(2020,10,1),10))
            );
            meals.add(
                new SetMeal("GoodChicken",30.0,"BlackChicken",
                        new Beer("Rain",4.0,
                                LocalDate.of(2020,10,1),10))
            );
    }

    private void use (Beer beer) {
        boolean success = false;
        LinkedList<Beer> dellist = new LinkedList<>();
        for (Beer value : beers) {
            if (value.isOverDate(LocalDate.now()))
                dellist.add(value);
        }
        beers.removeAll(dellist);
        for (int i=0;i<beers.size();i++)
            if (beers.get(i).equals(beer)) {
                beers.remove(i);
                success = true;
                break;
            }
        if (!success) throw new IngredientSortOutException(beer);
    }

    private void use (Juice juice) {
        boolean success = false;
        LinkedList<Juice> dellist = new LinkedList<>();
        for (Juice value : juices) {
            if (value.isOverDate(LocalDate.now()))
                dellist.add(value);
        }
        juices.removeAll(dellist);
        for (int i=0;i<juices.size();i++)
            if (juices.get(i).equals(juice)) {
                juices.remove(i);
                success = true;
                break;
            }
        if (!success) throw new IngredientSortOutException(juice);
    }


    @Override
    public void SellMeal(SetMeal meal) {
        Drinks drink = meal.getDrink();
        try {
            if (drink instanceof Beer)
                use((Beer) drink);
            else
                use((Juice) drink);
            amount += meal.getPrice();
        } catch (IngredientSortOutException e) {
            System.out.println("No "+e.getDrink());
        }
    }

    @Override
    public void GetGoods(LinkedList<Drinks> drinks) {
        try {
            for (Drinks drink : drinks){
                if (amount < drink.cost)
                    throw new OverdraftBalanceException(amount);
                else
                    amount -= drink.cost;
                if (drink instanceof Beer) {
                    beers.add((Beer) drink);
                    System.out.printf("Get a %s%n",drink.name);
                }
                else if (drink instanceof Juice){
                    juices.add((Juice) drink);
                    System.out.printf("Get a %s%n",drink.name);
                }
            }
        } catch (OverdraftBalanceException e) {
            System.out.printf("Only %.2f yuan. Cannot afford.%n",e.getAmount());
        }
    }

    public static void main(String[] args) {
        West2FriedChickenRestaurant west = new West2FriedChickenRestaurant();
        // 初始化中无 Version_Java.Drinks
        LinkedList<Drinks> initDrinks = new LinkedList<Drinks>();
        initDrinks.add(new Juice("Snow",4.0,
                LocalDate.of(2020,11,1)));
        initDrinks.add(new Beer("Sun",4.0,
                LocalDate.of(2020,11,6),10));
        initDrinks.add(new Beer("Rain",4.0,
                LocalDate.of(2020,11,6),10));
        west.GetGoods(initDrinks);
        west.SellMeal(new SetMeal("GoodChicken",30.0,"BlackChicken",
                new Beer("Rain",4.0,
                        LocalDate.of(2020,10,1),10)));
        west.SellMeal(new SetMeal("GoodChicken",30.0,"BlackChicken",
                new Beer("Rain",4.0,
                        LocalDate.of(2020,10,1),10)));
        west.GetGoods(initDrinks);
        west.GetGoods(initDrinks);
        west.GetGoods(initDrinks);
    }
}

File: src/main/java/Version_Kotlin/West2FriedChickenRestaurant.kt
package Version_Kotlin
import java.time.LocalDate
import java.util.*

class West2FriedChickenRestaurant : FriedChickenRestaurant {
    private var amount = 12.0 // 起始资金
    private val beers = LinkedList<Beer>()
    private val juices = LinkedList<Juice>()
    private val meals = LinkedList<SetMeal>()
    // Answer:
    // 使用 LinkedList 也就是 链表结构 添加 和 删除 更为高效

    init {
        meals.add(
                SetMeal("SuperChicken", 70.0, "WhiteChicken",
                        Juice("Snow", 4.0,
                                LocalDate.of(2020, 11, 1)))
        )
        meals.add(
                SetMeal("NiceChicken", 50.0, "GreenChicken",
                        Beer("Sun", 4.0,
                                LocalDate.of(2020, 10, 1), 10F))
        )
        meals.add(
                SetMeal("GoodChicken", 30.0, "BlackChicken",
                        Beer("Rain", 4.0,
                                LocalDate.of(2020, 10, 1), 10F))
        )
    }

    private fun use(beer: Beer) {
        var success = false
        val delList = LinkedList<Beer>()
        for (br in beers) {
            if (br.isOverDate(LocalDate.now())) {
                delList.add(br)
            }
        }
        beers.removeAll(delList)
        for (i in beers.indices)
            if (beers[i].equals(beer)) {
                beers.removeAt(i)
                success = true
                break
            }
        if (!success) throw IngredientSortOutException(beer)
    }

    private fun use(juice: Juice) {
        var success = false
        val delList = LinkedList<Juice>()
        for (je in juices) {
            if (je.isOverDate(LocalDate.now())) {
                delList.add(je)
            }
        }
        juices.removeAll(delList)
        for (je in juices) {
            if (je.isOverDate(LocalDate.now())) {
                println(je.toString())
                juices.remove(je)
            }
        }
        for (i in juices.indices)
            if (juices[i].equals(juice)) {
                juices.removeAt(i)
                success = true
                break
            }
        if (!success) throw IngredientSortOutException(juice)
    }

    override fun SellMeal(meal: SetMeal) {
        val drink: Drinks = meal.drink
        try {
            when (drink) {
                is Beer -> use(drink)
                is Juice -> use(drink)
            }
            println("Have sold one meal with " + drink.javaClass.toString())
            amount += meal.price
        } catch (e: IngredientSortOutException) {
            println("No " + e.getDrink())
        }
    }

    override fun GetGoods(drinks: LinkedList<Drinks>) {
        try {
            for (drink in drinks) {
                if (amount < drink.cost)
                    throw OverdraftBalanceException(amount)
                else amount -= drink.cost
                if (drink is Beer) {
                    beers.add(drink)
                    println(String.format("Get a %s", drink.name))
                } else if (drink is Juice) {
                    juices.add(drink)
                    println(String.format("Get a %s", drink.name))
                }
            }
        } catch (e: OverdraftBalanceException) {
            println(String.format("Only %.2f yuan. Cannot afford.", e.amount))
        }
    }


}